src/batched/sparse/KokkosBatched_JacobiPrec.hpp

```diff
@@ -0,0 +1,143 @@
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+#ifndef __KOKKOSBATCHED_JACOBIPREC_HPP__
+#define __KOKKOSBATCHED_JACOBIPREC_HPP__
+
+/// \author Kim Liegeois (knliege@sandia.gov)
+
+#include "KokkosBatched_HadamardProduct.hpp"
+
+namespace KokkosBatched {
+
+/// \brief Batched Jacobi Preconditioner:
+///
+/// \tparam ValuesViewType: Input type for the values of the diagonal
+
+template <class ValuesViewType>
+class JacobiPrec {
+ public:
+  using ScalarType = typename ValuesViewType::non_const_value_type;
+  using MagnitudeType =
+      typename Kokkos::Details::ArithTraits<ScalarType>::mag_type;
+
+ private:
+  mutable ValuesViewType diag_values;
+  int n_operators;
+  int n_rows;
+  int n_colums;
+  mutable bool computed_inverse = false;
+
+ public:
+  KOKKOS_INLINE_FUNCTION
+  JacobiPrec(const ValuesViewType &_diag_values) : diag_values(_diag_values) {
+    n_operators = _diag_values.extent(0);
+    n_rows      = _diag_values.extent(1);
+    n_colums    = n_rows;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  ~JacobiPrec() {}
+
+  template <typename MemberType, typename ArgMode>
+  KOKKOS_INLINE_FUNCTION void computeInverse(const MemberType &member) const {
+    auto one     = Kokkos::Details::ArithTraits<MagnitudeType>::one();
+    auto epsilon = Kokkos::Details::ArithTraits<MagnitudeType>::epsilon();
+    if (std::is_same<ArgMode, Mode::Serial>::value) {
+      for (int i = 0; i < n_operators; ++i)
+        for (int j = 0; j < n_colums; ++j) {
+          if (Kokkos::abs<ScalarType>(diag_values(i, j)) <= epsilon) {
```

Comment:
Why not checking when we do a team or teamvector algorithm?
How about having an integer value `int tooSmall` that keeps track of how many small values are detected and when such small value is detected replace it by one, or in practice do noting.
Then print a warning message to the user telling them how many small values where detected but not fail?